### ISSUE TITLE
Reindex from remote retry attempts

### ISSUE THREAD

**Issue by javanna** [elastic/elasticsearch#19395]
Currently, reindex from remote allows to configure one single host to send http requests to. In case of failure, there are cases (e.g. 429 response code for search) where we schedule a retry for the same request against the same node. We also have a backoff policy so that bulk failures get retried too.

I wonder if it makes sense to allow reindex from remote to either use the Sniffer and send requests to multiple nodes, or allow to configure multiple nodes, or both. In case multiple nodes are available, it makes much more sense to let the `RestClient` do the retries on a different node, then the current custom retry code would be less needed I think.

This issue is to discuss what makes sense to do in reindex from remote and whether we need to change anything there.

**Comment by javanna** [comment]
@nik9000 can you double check that what I said in the description is true and eventually correct me please?

**Comment by javanna** [comment]
Maybe also @elastic/es-clients would like to have a look. It is the first time that we have something in es that sends http requests, although this is more of a background process than an interactive thing, reason why maybe scheduling retries does make sense, for the sake of not making a long running operation (reindex) fail.

**Comment by polyfractal** [comment]
Hm, I don't see any reason why it *wouldn't* be a good idea to allow retrying against multiple nodes and/or sniffing the cluster state?  

Seems like it would help make reindex more robust without really introducing much extra complexity, just round-robin between the known nodes like the clients do.  It can keep the same backoff policy per-node, and only kill the job if all nodes are "marked dead" and their backoff timer isn't up?

**Comment by HonzaKral** [comment]
Agreed with Zach - it really should just round robin etc like regular clients, like the documented best practices. Including passing in multiple nodes via initial configuration and specifying custom timeout (indexing across data centers, slow network, weak cluster, ...). Also to follow same retry logic as the clients - retrying on `502`, `503` and `504` which is the current concensus.

We just need to make sure that anything that relies on environment (sniffing that won't work on nodes behind load balancer to give an example) can be turned off and is probably off by default.

**Comment by nik9000** [comment]
From a dependencies perspective I'd prefer not to sniff. I'd like to let
the user specify the hosts in the request.

As far as retries go, reindex only retries on too many requests right now.
We can change it but I don't want reindex from remote to drift too far from
normal reindex.

**Comment by Mpdreamz** [comment]
Echoing my agreement with zach :+1: 

I do agree a seperate backoff retry mechanism is also neccessary, whereas the clients are meant to exhaust retries fast (fail fast) in the case of reindex a backoff could really help users.

Why not expose these choices to the user more explicitly?

```json
hosts: ["x", "y"]
mode: "static|sniff"
backoff_interval_max: "5m",
backoff_interval: "1m"
````

`sniff` being default.

**Comment by javanna** [comment]
I feel like I need to clarify something: what I called backoff in the description of the issue, is to prevent retrying the same request on the same node forever, which is what reindex from remote currently does. Moving to the ordinary rest client behaviour would mean that a request would never be retried on the same node where it already failed (unless reindex schedules a retry like it currently does), but it would use load balancing and fail-over that come for free with the rest client. About configuring the backoff in the rest client, which is what Martijn was referring to , about when to send again a request to a node that previously failed, that configuration is not exposed to users at the moment in the java RestClient. I feel like this last bit would be part of a different discussion, but it seems like everybody agrees that reindex from remote shouldn't to too much magic under the hood.

@nik9000 maybe you can bring up the reasons why reindex currently retries when it gets back 429 rather than failing the reindex as a whole?

**Comment by nik9000** [comment]
> @nik9000 maybe you can bring up the reasons why reindex currently retries when it gets back 429 rather than failing the reindex as a whole?

The idea is to prevent failure when the cluster is overloaded. Reindex can stand to wait around for a bit if the cluster doesn't have room to add it to the queue. It does for bulk as well. It is part of reindex behaving like a management tool more than an interactive tool.

I'd love for users to be able to configure multiple hosts. We'll run them all through the whitelist, fail if they configure one that is forbidden, and then let the client round robin between them. That'll way we'll be able to handle if the node goes down. I mean, it won't be super pretty if one of the nodes holding the scroll context goes down, but at least we can handle the coordinating node going down.

I still don't like the idea of sniffing for reindex-from-remote. I think the nodes should be explicit so we don't surprise anyone. I'm not totally sure how sniffing would work with the whitelist.

**Comment by javanna** [comment]
I guess my last questions is: if we had multiple nodes, could we remove the manual retry on 429? I guess not cause for 4xx errors the client just returns them, no retry attempts on other nodes, and we couldn't manually retry on another host, cause you don't get to decide which host you send a request to.

**Comment by Mpdreamz** [comment]
We could not, thats what I meant. Theres two types of retries the one the client does (to provide failover if the node goes down), and a reindex specific backoff on 429 which makes a lot of sense to me too. 

I can imagine sniffing being icky with whitelists but we can still allow users to specify multiple hosts using a static connection pool and benefit from the clients failover capabilities.

**Comment by clintongormley** [comment]
> I still don't like the idea of sniffing for reindex-from-remote. I think the nodes should be explicit so we don't surprise anyone. I'm not totally sure how sniffing would work with the whitelist.

I think that we should at least take multiple nodes, like we do with the client.  I also think that sniffing would make this easier to configure than specifying multiple nodes -- you just specify one and get round-robining automatically.  And I'm still not convinced that we need a whitelist for reindex-from-remote.

**Comment by nihcas700** [comment]
Hi team, is this issue available to pick up ?
